### ISSUE TITLE
docker_container tries to import an obsolete docker-py which has been replaced by docker PyPi module

### ISSUE THREAD

**Issue by jean-christophe-manciot** [ansible/ansible#37200]
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
 docker_container

##### SUMMARY
Running the following playbook leads to the issue ```Failed to import docker-py - No module named ssl_match_hostname.```

##### ENVIRONMENT
* Ansible version: v2.4.3.0-1
* Operating System: Ubuntu 17.10
* docker-py: 1.10.6

##### STEPS TO REPRODUCE

Run the following playbook:
```
- name: Activate postgres container
  hosts: 127.0.0.1
  connection: local
  gather_facts: false
  tasks:
        - docker_container:
            name: postgres
            state: started
            restart_policy: unless-stopped
            image: postgres:9.6
            volumes:
              - "/var/lib/pgdocker"
            env:
              POSTGRES_USER: 'admin'
              POSTGRES_PASSWORD: 'password'
              POSTGRES_DB: '/var/lib/postgresql'
              PGDATA: "/var/lib/postgresql/data/pgdata"
          when: pg_hostname is not defined or pg_hostname == ''
          register: postgres_container_activate
```

##### EXPECTED RESULTS

Postgres container should be activated.

##### ACTUAL RESULTS

```
< TASK [docker_container] >
 -------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

task path: /media/actionmystique/SAMSUNG5-Shared/home/actionmystique/src/Ansible/test-docker_container/postgres.yml:7
Using module file /media/actionmystique/SAMSUNG5-Shared/home/actionmystique/src/Ansible/git-ansible/lib/ansible/modules/cloud/docker/docker_container.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python && sleep 0'
The full traceback is:
  File "/tmp/ansible_W0cPuK/ansible_module_docker_container.py", line 683, in <module>
    from docker import utils
  File "/usr/local/lib/python2.7/dist-packages/docker/__init__.py", line 6, in <module>
    from .client import Client, AutoVersionClient, from_env # flake8: noqa
  File "/usr/local/lib/python2.7/dist-packages/docker/client.py", line 11, in <module>
    from . import api
  File "/usr/local/lib/python2.7/dist-packages/docker/api/__init__.py", line 2, in <module>
    from .build import BuildApiMixin
  File "/usr/local/lib/python2.7/dist-packages/docker/api/build.py", line 6, in <module>
    from .. import constants

fatal: [127.0.0.1]: FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "api_version": null, 
            "auto_remove": false, 
            "blkio_weight": null, 
            "cacert_path": null, 
            "capabilities": null, 
            "cert_path": null, 
            "cleanup": false, 
            "command": null, 
            "cpu_period": null, 
            "cpu_quota": null, 
            "cpu_shares": null, 
            "cpuset_cpus": null, 
            "cpuset_mems": null, 
            "debug": false, 
            "detach": true, 
            "devices": null, 
            "dns_opts": null, 
            "dns_search_domains": null, 
            "dns_servers": null, 
            "docker_host": null, 
            "entrypoint": null, 
            "env": {
                "PGDATA": "/var/lib/postgresql/data/pgdata", 
                "POSTGRES_DB": "/var/lib/postgresql", 
                "POSTGRES_PASSWORD": "password", 
                "POSTGRES_USER": "admin"
            }, 
            "env_file": null, 
            "etc_hosts": null, 
            "exposed_ports": null, 
            "filter_logger": false, 
            "force_kill": false, 
            "groups": null, 
            "hostname": null, 
            "ignore_image": false, 
            "image": "postgres:9.6", 
            "interactive": false, 
            "ipc_mode": null, 
            "keep_volumes": true, 
            "kernel_memory": null, 
            "key_path": null, 
            "kill_signal": null, 
            "labels": null, 
            "links": null, 
            "log_driver": null, 
            "log_options": null, 
            "mac_address": null, 
            "memory": "0", 
            "memory_reservation": null, 
            "memory_swap": null, 
            "memory_swappiness": null, 
            "name": "postgres", 
            "network_mode": null, 
            "networks": null, 
            "oom_killer": null, 
            "oom_score_adj": null, 
            "paused": false, 
            "pid_mode": null, 
            "privileged": false, 
            "published_ports": null, 
            "pull": false, 
            "purge_networks": false, 
            "read_only": false, 
            "recreate": false, 
            "restart": false, 
            "restart_policy": "unless-stopped", 
            "restart_retries": null, 
            "security_opts": null, 
            "shm_size": null, 
            "ssl_version": null, 
            "state": "started", 
            "stop_signal": null, 
            "stop_timeout": null, 
            "sysctls": null, 
            "timeout": null, 
            "tls": null, 
            "tls_hostname": null, 
            "tls_verify": null, 
            "tmpfs": null, 
            "trust_image_content": false, 
            "tty": false, 
            "ulimits": null, 
            "user": null, 
            "uts": null, 
            "volume_driver": null, 
            "volumes": [
                "/var/lib/pgdocker"
            ], 
            "volumes_from": null, 
            "working_dir": null
        }
    }, 
    "msg": "Failed to import docker-py - No module named ssl_match_hostname. Try `pip install docker-py`"
}
```

##### ADDITIONAL INFORMATION
```
# pip install docker-py
Requirement already satisfied: docker-py in /usr/local/lib/python2.7/dist-packages
Requirement already satisfied: docker-pycreds>=0.2.1 in /usr/local/lib/python2.7/dist-packages (from docker-py)
Requirement already satisfied: backports.ssl-match-hostname>=3.5; python_version < "3.5" in /usr/local/lib/python2.7/dist-packages (from docker-py)
Requirement already satisfied: ipaddress>=1.0.16; python_version < "3.3" in /usr/local/lib/python2.7/dist-packages (from docker-py)
Requirement already satisfied: requests!=2.11.0,>=2.5.2 in /usr/local/lib/python2.7/dist-packages (from docker-py)
Requirement already satisfied: six>=1.4.0 in /usr/local/lib/python2.7/dist-packages (from docker-py)
Requirement already satisfied: websocket-client>=0.32.0 in /usr/local/lib/python2.7/dist-packages (from docker-py)
Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python2.7/dist-packages (from requests!=2.11.0,>=2.5.2->docker-py)
Requirement already satisfied: chardet<3.1.0,>=3.0.2 in /usr/local/lib/python2.7/dist-packages (from requests!=2.11.0,>=2.5.2->docker-py)
Requirement already satisfied: idna<2.7,>=2.5 in /usr/local/lib/python2.7/dist-packages (from requests!=2.11.0,>=2.5.2->docker-py)
Requirement already satisfied: urllib3<1.23,>=1.21.1 in /usr/local/lib/python2.7/dist-packages (from requests!=2.11.0,>=2.5.2->docker-py)
```

**Comment by jean-christophe-manciot** [comment]
Exact same issue on the latest unstable version:
```
ansible 2.6.0 (devel 0130490faa) last updated 2018/03/08 17:34:08 (GMT +200)
```

**Comment by ansibot** [comment]
cc @chouseknecht @cove @dusdanig @jctanner @joshuaconner @kassiansun @softzilla @zfil
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sivel** [comment]
This is not a direct issue of ansible.  The `docker-py` python module requires `backports.ssl-match-hostname` as a dependency and must be installed.  A typical `pip install docker-py` should install something similar to:

```
Collecting backports.ssl-match-hostname>=3.5; python_version < "3.5" (from docker-py)
  Downloading backports.ssl_match_hostname-3.5.0.1.tar.gz
```

I see from above that your pip install is reporting that the dependency is already resolved, however docker-py seems to be unable to find it.

Try uninstalling and reinstalling `backports.ssl-match-hostname` or force installing with `-I`.  We have seen issues where other `backports` modules may interfere with each other (see https://github.com/peterjc/backports.lzma/pull/32)

If you continue experiencing issues, please open an issue with the upstream docker-py project or related dependency.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Comment by jean-christophe-manciot** [comment]
@sivel 
I have 2 comments:
- have you really tried to run the aforementioned playbook on your host? You'll most probably experience the same bug.
- I cannot have any backports module incompatibilities:
```
# pip list --format columns | grep -E "^backports\..* "
backports.ssl-match-hostname 3.5.0.1          
```

**Comment by jean-christophe-manciot** [comment]
- I also tried all **9** versions of ```docker-py``` from 1.10.6 down to 1.10.0: they all yield to the same issue.
- Same remark about all **3** last releases of ```backports.ssl-match-hostname``` between and including 3.5.0.1 --> 3.4.0.1.

I doubt that there is an issue directly from these packages.

**Comment by jean-christophe-manciot** [comment]
@sivel 
As indicated by @TomasTomecek when I reported this issue to the [docker-py project](https://github.com/docker/docker-py/issues/1947#issuecomment-371832526), the former PyPi module```docker-py```  **is not maintained anymore**.
As indicated by the [official documentation](https://github.com/docker/docker-py), it must now be install with ```pip install docker``` (also cf. [docker on PyPi](https://pypi.python.org/pypi/docker/)).

However, even when uninstalling docker-py and installing docker python module, the issue is still there.
The conclusion is that **the ansible docker_container module must import docker instead of docker-py**:
```
# pip show docker
Name: docker
Version: 3.1.1
Summary: A Python library for the Docker Engine API.
Home-page: https://github.com/docker/docker-py
Author: Joffrey F
Author-email: joffrey@docker.com
License: Apache License 2.0
Location: /usr/local/lib/python2.7/dist-packages
Requires: backports.ssl-match-hostname, ipaddress, six, docker-pycreds, requests, websocket-client
```

**Comment by sivel** [comment]
The docker modules within ansible already contain functionality to properly import `docker` vs `docker-py`.  Due note however that you cannot have both `docker` and `docker-py` installed at the same time, as they utilize the same namespace.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.freenode.net
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Comment by jean-christophe-manciot** [comment]
No, I have only docker module installed and the issue is still there:
```
# pip install -I docker
Collecting docker
  Using cached docker-3.1.1-py2.py3-none-any.whl
Collecting backports.ssl-match-hostname>=3.5; python_version < "3.5" (from docker)
Collecting ipaddress>=1.0.16; python_version < "3.3" (from docker)
Collecting six>=1.4.0 (from docker)
  Using cached six-1.11.0-py2.py3-none-any.whl
Collecting docker-pycreds>=0.2.2 (from docker)
  Using cached docker_pycreds-0.2.2-py2.py3-none-any.whl
Collecting requests!=2.18.0,>=2.14.2 (from docker)
  Using cached requests-2.18.4-py2.py3-none-any.whl
Collecting websocket-client>=0.32.0 (from docker)
  Using cached websocket_client-0.47.0-py2.py3-none-any.whl
Collecting certifi>=2017.4.17 (from requests!=2.18.0,>=2.14.2->docker)
  Using cached certifi-2018.1.18-py2.py3-none-any.whl
Collecting chardet<3.1.0,>=3.0.2 (from requests!=2.18.0,>=2.14.2->docker)
  Using cached chardet-3.0.4-py2.py3-none-any.whl
Collecting idna<2.7,>=2.5 (from requests!=2.18.0,>=2.14.2->docker)
  Using cached idna-2.6-py2.py3-none-any.whl
Collecting urllib3<1.23,>=1.21.1 (from requests!=2.18.0,>=2.14.2->docker)
  Using cached urllib3-1.22-py2.py3-none-any.whl
Installing collected packages: backports.ssl-match-hostname, ipaddress, six, docker-pycreds, certifi, chardet, idna, urllib3, requests, websocket-client, docker
Successfully installed backports.ssl-match-hostname-3.5.0.1 certifi-2018.1.18 chardet-3.0.4 docker-3.1.1 docker-pycreds-0.2.2 idna-2.6 ipaddress-1.0.19 requests-2.18.4 six-1.11.0 urllib3-1.22 websocket-client-0.47.0

r# pip list --format columns | grep -E "^docker.* "
docker                       3.1.1            
docker-pycreds               0.2.2         
   
# ansible-playbook -vvvv postgres.yml 
ansible-playbook 2.4.3.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/actionmystique/src/Ansible/git-yang-networkop/ansible-101/library']
  ansible python module location = /usr/lib/python2.7/dist-packages/ansible
  executable location = /usr/bin/ansible-playbook
  python version = 2.7.14+ (default, Feb  7 2018, 08:39:52) [GCC 7.3.0]
Using /etc/ansible/ansible.cfg as config file
setting up inventory plugins
Parsed /etc/ansible/hosts inventory source with ini plugin
 [WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

Loading callback plugin default of type stdout, v2.0 from /usr/lib/python2.7/dist-packages/ansible/plugins/callback/default.pyc
 ________________________
< PLAYBOOK: postgres.yml >
 ------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

1 plays in postgres.yml
 ____________________________________
< PLAY [Activate postgres container] >
 ------------------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

META: ran handlers
 _________________________
< TASK [docker_container] >
 -------------------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

task path: /media/actionmystique/SAMSUNG5-Shared/home/actionmystique/src/Ansible/test-docker_container/postgres.yml:7
Using module file /usr/lib/python2.7/dist-packages/ansible/modules/cloud/docker/docker_container.py
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: root
<127.0.0.1> EXEC /bin/sh -c '/usr/bin/python && sleep 0'
The full traceback is:
  File "/tmp/ansible_6KdZPJ/ansible_module_docker_container.py", line 683, in <module>
    from docker import utils
  File "/usr/local/lib/python2.7/dist-packages/docker/__init__.py", line 2, in <module>
    from .api import APIClient
  File "/usr/local/lib/python2.7/dist-packages/docker/api/__init__.py", line 2, in <module>
    from .client import APIClient
  File "/usr/local/lib/python2.7/dist-packages/docker/api/client.py", line 10, in <module>
    from .build import BuildApiMixin
  File "/usr/local/lib/python2.7/dist-packages/docker/api/build.py", line 5, in <module>
    from .. import auth
  File "/usr/local/lib/python2.7/dist-packages/docker/auth.py", line 8, in <module>
    from . import errors

fatal: [127.0.0.1]: FAILED! => {
    "changed": false, 
    "invocation": {
        "module_args": {
            "api_version": null, 
            "auto_remove": false, 
            "blkio_weight": null, 
            "cacert_path": null, 
            "capabilities": null, 
            "cert_path": null, 
            "cleanup": false, 
            "command": null, 
            "cpu_period": null, 
            "cpu_quota": null, 
            "cpu_shares": null, 
            "cpuset_cpus": null, 
            "cpuset_mems": null, 
            "debug": false, 
            "detach": true, 
            "devices": null, 
            "dns_opts": null, 
            "dns_search_domains": null, 
            "dns_servers": null, 
            "docker_host": null, 
            "entrypoint": null, 
            "env": {
                "PGDATA": "/var/lib/postgresql/data/pgdata", 
                "POSTGRES_DB": "/var/lib/postgresql", 
                "POSTGRES_PASSWORD": "password", 
                "POSTGRES_USER": "admin"
            }, 
            "env_file": null, 
            "etc_hosts": null, 
            "exposed_ports": null, 
            "filter_logger": false, 
            "force_kill": false, 
            "groups": null, 
            "hostname": null, 
            "ignore_image": false, 
            "image": "postgres:9.6", 
            "interactive": false, 
            "ipc_mode": null, 
            "keep_volumes": true, 
            "kernel_memory": null, 
            "key_path": null, 
            "kill_signal": null, 
            "labels": null, 
            "links": null, 
            "log_driver": null, 
            "log_options": null, 
            "mac_address": null, 
            "memory": "0", 
            "memory_reservation": null, 
            "memory_swap": null, 
            "memory_swappiness": null, 
            "name": "postgres", 
            "network_mode": null, 
            "networks": null, 
            "oom_killer": null, 
            "oom_score_adj": null, 
            "paused": false, 
            "pid_mode": null, 
            "privileged": false, 
            "published_ports": null, 
            "pull": false, 
            "purge_networks": false, 
            "read_only": false, 
            "recreate": false, 
            "restart": false, 
            "restart_policy": "unless-stopped", 
            "restart_retries": null, 
            "security_opts": null, 
            "shm_size": null, 
            "ssl_version": null, 
            "state": "started", 
            "stop_signal": null, 
            "stop_timeout": null, 
            "sysctls": null, 
            "timeout": null, 
            "tls": null, 
            "tls_hostname": null, 
            "tls_verify": null, 
            "tmpfs": null, 
            "trust_image_content": false, 
            "tty": false, 
            "ulimits": null, 
            "user": null, 
            "uts": null, 
            "volume_driver": null, 
            "volumes": [
                "/var/lib/pgdocker"
            ], 
            "volumes_from": null, 
            "working_dir": null
        }
    }, 
    "msg": "Failed to import docker-py - No module named ssl_match_hostname. Try `pip install docker-py`"
}
```

**Comment by Vye** [comment]
I'm also affected by this. I've followed the steps outlined here and still get the same error as @jean-christophe-manciot. Can we get this re-opened @sivel? 

```
ansible-playbook 2.4.3.0
  config file = /Users/vye/repos/kong/ansible.cfg
  configured module search path = ['/Users/vye/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ansible
  executable location = /Library/Frameworks/Python.framework/Versions/3.6/bin/ansible-playbook
  python version = 3.6.0 (v3.6.0:41df79263a11, Dec 22 2016, 17:23:13) [GCC 4.2.1 (Apple Inc. build 5666) (dot 3)]
```

```
pip3 list --format columns | grep -E "^docker.* "
docker                   3.1.1
docker-compose           1.9.0
docker-pycreds           0.2.2
dockerpty                0.4.1

```

```
TASK [Create container] ***********************************************************************************************************************************************************************************************************
task path: /Users/vye/repos/kong/hooks/ec2.stage.start.yml:91
Thursday 15 March 2018  13:09:22 -0700 (0:00:00.040)       0:00:02.184 ********
Using module_utils file /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ansible/module_utils/six/__init__.py
Using module_utils file /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ansible/module_utils/docker_common.py
Using module_utils file /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ansible/module_utils/basic.py
Using module_utils file /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ansible/module_utils/parsing/convert_bool.py
Using module_utils file /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ansible/module_utils/parsing/__init__.py
Using module_utils file /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ansible/module_utils/_text.py
Using module_utils file /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ansible/module_utils/pycompat24.py
Using module file /Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/ansible/modules/cloud/docker/docker_container.py
<localhost> ESTABLISH LOCAL CONNECTION FOR USER: vye
<localhost> EXEC /bin/sh -c 'echo ~ && sleep 0'
<localhost> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/vye/.ansible/tmp/ansible-tmp-1521144562.748009-176987348925506 `" && echo ansible-tmp-1521144562.748009-176987348925506="` echo /Users/vye/.ansible/tmp/ansible-tmp-1521144562.748009-176987348925506 `" ) && sleep 0'
<localhost> PUT /var/folders/ch/_8p0ymrd043_ycdxzsjtq47nk192r3/T/tmpugu4jpet TO /Users/vye/.ansible/tmp/ansible-tmp-1521144562.748009-176987348925506/docker_container.py
<localhost> EXEC /bin/sh -c 'chmod u+x /Users/vye/.ansible/tmp/ansible-tmp-1521144562.748009-176987348925506/ /Users/vye/.ansible/tmp/ansible-tmp-1521144562.748009-176987348925506/docker_container.py && sleep 0'
<localhost> EXEC /bin/sh -c 'AWS_PROFILE=dev.pipeline /usr/bin/python /Users/vye/.ansible/tmp/ansible-tmp-1521144562.748009-176987348925506/docker_container.py; rm -rf "/Users/vye/.ansible/tmp/ansible-tmp-1521144562.748009-176987348925506/" > /dev/null 2>&1 && sleep 0'
The full traceback is:
  File "/var/folders/ch/_8p0ymrd043_ycdxzsjtq47nk192r3/T/ansible_AogKVg/ansible_module_docker_container.py", line 683, in <module>
    from docker import utils
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/docker/__init__.py", line 2, in <module>
    from .api import APIClient
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/docker/api/__init__.py", line 2, in <module>
    from .client import APIClient
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/docker/api/client.py", line 10, in <module>
    from .build import BuildApiMixin
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/docker/api/build.py", line 5, in <module>
    from .. import auth
  File "/Library/Frameworks/Python.framework/Versions/3.6/lib/python3.6/site-packages/docker/auth.py", line 8, in <module>
    from . import errors

fatal: [localhost]: FAILED! => {
    "changed": false,
    "invocation": {
        "module_args": {
            "api_version": null,
            "auto_remove": false,
            "blkio_weight": null,
            "cacert_path": null,
            "capabilities": null,
            "cert_path": null,
            "cleanup": false,
            "command": null,
            "cpu_period": null,
            "cpu_quota": null,
            "cpu_shares": null,
            "cpuset_cpus": null,
            "cpuset_mems": null,
            "debug": false,
            "detach": true,
            "devices": null,
            "dns_opts": null,
            "dns_search_domains": null,
            "dns_servers": null,
            "docker_host": null,
            "entrypoint": null,
            "env": null,
            "env_file": null,
            "etc_hosts": null,
            "exposed_ports": null,
            "filter_logger": false,
            "force_kill": false,
            "groups": null,
            "hostname": null,
            "ignore_image": false,
            "image": "builder",
            "interactive": false,
            "ipc_mode": null,
            "keep_volumes": true,
            "kernel_memory": null,
            "key_path": null,
            "kill_signal": null,
            "labels": null,
            "links": null,
            "log_driver": null,
            "log_options": null,
            "mac_address": null,
            "memory": "0",
            "memory_reservation": null,
            "memory_swap": null,
            "memory_swappiness": null,
            "name": "builder-1",
            "network_mode": null,
            "networks": null,
            "oom_killer": null,
            "oom_score_adj": null,
            "paused": false,
            "pid_mode": null,
            "privileged": false,
            "published_ports": null,
            "pull": false,
            "purge_networks": false,
            "read_only": false,
            "recreate": false,
            "restart": false,
            "restart_policy": null,
            "restart_retries": null,
            "security_opts": null,
            "shm_size": null,
            "ssl_version": null,
            "state": "started",
            "stop_signal": null,
            "stop_timeout": null,
            "sysctls": null,
            "timeout": null,
            "tls": null,
            "tls_hostname": null,
            "tls_verify": null,
            "tmpfs": null,
            "trust_image_content": false,
            "tty": false,
            "ulimits": null,
            "user": null,
            "uts": null,
            "volume_driver": null,
            "volumes": null,
            "volumes_from": null,
            "working_dir": null
        }
    }
}

MSG:

Failed to import docker-py - No module named ssl_match_hostname. Try `pip install docker-py`

```

**Comment by TomasTomecek** [comment]
The error message is quite clear: `No module named ssl_match_hostname`.

Do you have it installed?
```
$ pip2 freeze | grep match
backports.ssl-match-hostname==3.5.0.1
```

What I find really interesting is that the package makes only sense for python2, since for python3 it is part of stdlib. I can clearly see you are using python 3.6, is the error message misleading?

**Comment by jean-christophe-manciot** [comment]
@TomasTomecek 
```
# pip2 freeze | grep match
backports.ssl-match-hostname==3.5.0.1
```
What I find really interesting is that I have posted multiple bug reports about this sole issue, and ansible people keep closing them for no real technical reason.
As if red hat open-sourced AWX for the headlines but don't really want people to be able to use it in place of Tower. To clarify this last sentence, the playbook used in this issue is part of AWX installer.

**Comment by TomasTomecek** [comment]
> As if red hat open-sourced AWX for the headlines but don't really want people to be able to use it in place of Tower. 

You are confusing two things: Red Hat Ansible Tower as a commercial product, AWX as an open source project. With the first one you pay for support, the second one is an open source project with support provided on best-effort basis. As the license states:
```
  15. Disclaimer of Warranty.

  THERE IS NO WARRANTY FOR THE PROGRAM, TO THE EXTENT PERMITTED BY
APPLICABLE LAW.  EXCEPT WHEN OTHERWISE STATED IN WRITING THE COPYRIGHT
HOLDERS AND/OR OTHER PARTIES PROVIDE THE PROGRAM "AS IS" WITHOUT WARRANTY
OF ANY KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING, BUT NOT LIMITED TO,
THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
PURPOSE.  THE ENTIRE RISK AS TO THE QUALITY AND PERFORMANCE OF THE PROGRAM
IS WITH YOU.  SHOULD THE PROGRAM PROVE DEFECTIVE, YOU ASSUME THE COST OF
ALL NECESSARY SERVICING, REPAIR OR CORRECTION.
```

We are sorry that you are hitting this issue but there is no engineer with free time to work on this at the moment. Including me, I would love to resolve it, but I just don't have time to do it.

**Comment by TomasTomecek** [comment]
What happens when you import docker-py on the target environment directly from python?
```
$ python3 -c "import docker"
```

**Comment by jean-christophe-manciot** [comment]
I am not confusing anything. I am fully aware that AWX is the upstream open-source project for Tower.
I suppose you mean python2, right ;-) ?
```
# export PYTHONPATH=${PYTHONPATH}/usr/local/lib/python2.7/dist-package
# echo "PYTHONPATH ${PYTHONPATH}"
PYTHONPATH /usr/local/lib/python2.7/dist-packages
# python2 -c "import docker"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/usr/local/lib/python2.7/dist-packages/docker/__init__.py", line 2, in <module>
    from .api import APIClient
  File "/usr/local/lib/python2.7/dist-packages/docker/api/__init__.py", line 2, in <module>
    from .client import APIClient
  File "/usr/local/lib/python2.7/dist-packages/docker/api/client.py", line 10, in <module>
    from .build import BuildApiMixin
  File "/usr/local/lib/python2.7/dist-packages/docker/api/build.py", line 5, in <module>
    from .. import auth
  File "/usr/local/lib/python2.7/dist-packages/docker/auth.py", line 9, in <module>
    from .utils import config
  File "/usr/local/lib/python2.7/dist-packages/docker/utils/__init__.py", line 2, in <module>
    from .build import tar, exclude_paths
  File "/usr/local/lib/python2.7/dist-packages/docker/utils/build.py", line 7, in <module>
    from .utils import create_archive
  File "/usr/local/lib/python2.7/dist-packages/docker/utils/utils.py", line 16, in <module>
    from .. import tls
  File "/usr/local/lib/python2.7/dist-packages/docker/tls.py", line 5, in <module>
    from .transport import SSLAdapter
  File "/usr/local/lib/python2.7/dist-packages/docker/transport/__init__.py", line 3, in <module>
    from .ssladapter import SSLAdapter
  File "/usr/local/lib/python2.7/dist-packages/docker/transport/ssladapter.py", line 21, in <module>
    from backports.ssl_match_hostname import match_hostname
ImportError: No module named ssl_match_hostname
```
despite:
```
# ll /usr/local/lib/python2.7/dist-packages/backports.ssl_match_hostname-3.5.0.1.dist-info/
total 72
drwxr-xr-x   2 root root  4096 Mar  9 17:19 ./
drwxr-xr-x 720 root root 36864 Mar  9 17:19 ../
-rw-r--r--   1 root root  2635 Mar  9 17:19 DESCRIPTION.rst
-rw-r--r--   1 root root     4 Mar  9 17:19 INSTALLER
-rw-r--r--   1 root root  3487 Mar  9 17:19 METADATA
-rw-r--r--   1 root root   965 Mar  9 17:19 metadata.json
-rw-r--r--   1 root root   989 Mar  9 17:19 RECORD
-rw-r--r--   1 root root    10 Mar  9 17:19 top_level.txt
-rw-r--r--   1 root root    93 Mar  9 17:19 WHEEL
```
So, **python docker module cannot import the installed backports.ssl_match_hostname module**.

**Comment by jean-christophe-manciot** [comment]
This issue regarding the docker python module has been [reported ~ 1 year ago](https://github.com/docker/docker-py/issues/1502) and is still open/unsolved.

**Comment by TomasTomecek** [comment]
I got confused by the other report which is happening with python 3.6.

It seems that [you figured](https://github.com/docker/docker-py/issues/1502#issuecomment-373907682) it out yourself. See? That power of open source! =)

**Comment by jean-christophe-manciot** [comment]
Yes, except that:
- the debian version of python-docker is old when compared to docker pip module: 2.5.1 vs 3.1.1
- I'm experiencing [another issue](https://groups.google.com/d/msg/docker-dev/W72oJ4vrYUY/qLAl2ij5EQAJ) which seems related to an incompatibility between python-docker and docker-ce itself.

**Comment by jean-christophe-manciot** [comment]
I have found a [workaround](https://github.com/docker/docker-py/issues/1502#issuecomment-374172190) allowing me to finally be able to run the playbook described in this thread and install awx.

**Comment by afrazkhan** [comment]
Can you confirm that the official "fix" is either _not_ to use the `docker` module from Pip, and instead install "python-docker" package, or to use virtualenv?

I can confirm that the former does work, but just want to check that this is issue will not be fixed, and that is really the way to go.

**Comment by jean-christophe-manciot** [comment]
Here is my final successfully tested setup  with the latest packages (```python-docker``` is kind of old) without the need to use a virtualenv:
- docker-ce                          deb from docker PPA
- docker                             pip2 from PyPi
- docker-compose        pip2 from PyPi
- requests                           pip2 from PyPi (==2.18.1 and not 2.18.4)

The last line is the crucial part. cf. [my last post](https://github.com/docker/docker-py/issues/1502#issuecomment-374565304).

**Comment by vpopescu** [comment]
I ran into this as well (and it's late October now). I wish someone would write a definitive fix... the problem with this fix is: "azure-keyvault 1.0.0a1 has requirement requests>=2.18.4, but you'll have requests 2.18.1 which is incompatible. "
